File: frontend/utils/deviceFingerprint.ts
export interface ParsedUserAgent {
  osName: string
  browserName: string
  browserVersion: string
  platform: string
}

export interface DeviceFingerprint {
  deviceId: string
  osName: string
  browserName: string
  browserVersion: string
  platform: string
  deviceLabel: string
  headers: Record<string, string>
}

interface DeviceFingerprintOptions {
  deviceIdKey?: string
  deviceIdPrefix?: string
}

const GENERIC_DEVICE_NAMES = new Set([
  '',
  'web',
  'desktop',
  'android',
  'ios',
  'unknown',
  'current-browser',
  'browser'
])

function extractVersion(userAgent: string, patterns: RegExp[]): string {
  for (const pattern of patterns) {
    const match = userAgent.match(pattern)
    if (match?.[1]) return match[1]
  }
  return ''
}

function detectOsName(userAgent: string): string {
  const ua = userAgent.toLowerCase()
  if (ua.includes('windows nt') || ua.includes('windows')) return 'Windows'
  if (ua.includes('iphone') || ua.includes('ipad') || ua.includes('ipod') || ua.includes('ios')) return 'iOS'
  if (ua.includes('android')) return 'Android'
  if (ua.includes('macintosh') || ua.includes('mac os x')) return 'macOS'
  if (ua.includes('cros')) return 'ChromeOS'
  if (ua.includes('linux')) return 'Linux'
  return 'Unknown OS'
}

function detectBrowserNameAndVersion(userAgent: string): { browserName: string; browserVersion: string } {
  const ua = userAgent.toLowerCase()

  if (ua.includes('edg/') || ua.includes('edge/') || ua.includes('edga/') || ua.includes('edgios/')) {
    return { browserName: 'Edge', browserVersion: extractVersion(userAgent, [/(?:Edg|Edge|EdgA|EdgiOS)\/([\d.]+)/i]) }
  }
  if (ua.includes('opr/') || ua.includes('opera')) {
    return { browserName: 'Opera', browserVersion: extractVersion(userAgent, [/(?:OPR|Opera)\/([\d.]+)/i]) }
  }
  if (ua.includes('samsungbrowser/')) {
    return { browserName: 'Samsung Internet', browserVersion: extractVersion(userAgent, [/SamsungBrowser\/([\d.]+)/i]) }
  }
  if (ua.includes('firefox/') || ua.includes('fxios/')) {
    return { browserName: 'Firefox', browserVersion: extractVersion(userAgent, [/(?:Firefox|FxiOS)\/([\d.]+)/i]) }
  }
  if (ua.includes('micromessenger/')) {
    return { browserName: 'WeChat', browserVersion: extractVersion(userAgent, [/MicroMessenger\/([\d.]+)/i]) }
  }
  if (ua.includes('ucbrowser/')) {
    return { browserName: 'UC Browser', browserVersion: extractVersion(userAgent, [/UCBrowser\/([\d.]+)/i]) }
  }
  if (ua.includes('qqbrowser/')) {
    return { browserName: 'QQ Browser', browserVersion: extractVersion(userAgent, [/QQBrowser\/([\d.]+)/i]) }
  }
  if (ua.includes('msie ') || ua.includes('trident/')) {
    return { browserName: 'Internet Explorer', browserVersion: extractVersion(userAgent, [/MSIE\s([\d.]+)/i, /rv:([\d.]+)/i]) }
  }
  if (ua.includes('safari/') && !ua.includes('chrome/') && !ua.includes('crios/')) {
    return { browserName: 'Safari', browserVersion: extractVersion(userAgent, [/Version\/([\d.]+)/i]) }
  }
  if (ua.includes('chrome/') || ua.includes('crios/')) {
    return { browserName: 'Chrome', browserVersion: extractVersion(userAgent, [/(?:Chrome|CriOS)\/([\d.]+)/i]) }
  }

  return { browserName: 'Unknown Browser', browserVersion: '' }
}

function inferPlatformFromOs(osName: string): string {
  if (osName === 'iOS') return 'ios'
  if (osName === 'Android') return 'android'
  if (['Windows', 'macOS', 'Linux', 'ChromeOS'].includes(osName)) return 'desktop'
  return 'web'
}

export function parseUserAgent(userAgent?: string): ParsedUserAgent {
  const ua = String(userAgent || '')
  const osName = detectOsName(ua)
  const { browserName, browserVersion } = detectBrowserNameAndVersion(ua)
  return {
    osName,
    browserName,
    browserVersion,
    platform: inferPlatformFromOs(osName)
  }
}

export function buildDeviceLabel(osName?: string, browserName?: string): string {
  const safeOs = String(osName || '').trim() || 'Unknown OS'
  const safeBrowser = String(browserName || '').trim() || 'Unknown Browser'
  return `${safeOs} · ${safeBrowser}`
}

export function isGenericDeviceName(value?: string): boolean {
  const raw = String(value || '').trim()
  const lowered = raw.toLowerCase()
  if (GENERIC_DEVICE_NAMES.has(lowered)) return true
  if (/^(desktop|web|android|ios|unknown)\s*[·-]\s*[a-z]{2}(?:-[a-z]{2})?$/.test(lowered)) return true
  return lowered.endsWith(' browser')
}

export function normalizeDeviceName(value: string | undefined, parsed: ParsedUserAgent): string {
  const raw = String(value || '').trim()
  if (raw && !isGenericDeviceName(raw)) return raw
  return buildDeviceLabel(parsed.osName, parsed.browserName)
}

function ensureDeviceId(key: string, prefix: string): string {
  if (!import.meta.client) return ''
  let value = localStorage.getItem(key) || ''
  if (!value) {
    const randomPart = Math.random().toString(36).slice(2, 10)
    value = typeof crypto !== 'undefined' && 'randomUUID' in crypto
      ? crypto.randomUUID()
      : `${prefix}-${Date.now()}-${randomPart}`
    localStorage.setItem(key, value)
  }
  return value
}

export function getClientDeviceFingerprint(options?: DeviceFingerprintOptions): DeviceFingerprint {
  const deviceIdKey = options?.deviceIdKey || 'tg_device_id'
  const deviceIdPrefix = options?.deviceIdPrefix || 'dev'
  const userAgent = import.meta.client ? (navigator.userAgent || '') : ''
  const parsed = parseUserAgent(userAgent)
  const deviceId = ensureDeviceId(deviceIdKey, deviceIdPrefix)
  const deviceLabel = buildDeviceLabel(parsed.osName, parsed.browserName)
  const headers: Record<string, string> = {}
  if (deviceId) headers['X-Device-Id'] = deviceId
  if (parsed.platform) headers['X-Platform'] = parsed.platform
  if (deviceLabel) headers['X-Device-Name'] = deviceLabel.slice(0, 120)

  return {
    deviceId,
    osName: parsed.osName,
    browserName: parsed.browserName,
    browserVersion: parsed.browserVersion,
    platform: parsed.platform,
    deviceLabel,
    headers
  }
}

export function parseUserAgentToLabel(userAgent?: string): string {
  const parsed = parseUserAgent(userAgent)
  return buildDeviceLabel(parsed.osName, parsed.browserName)
}

File: tg_imagebed/device_fingerprint.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Device and browser fingerprint helpers."""

from __future__ import annotations

import re
from typing import Dict


_GENERIC_DEVICE_NAMES = {
    "",
    "web",
    "desktop",
    "android",
    "ios",
    "unknown",
    "current-browser",
    "browser",
}


def _extract_version(user_agent: str, patterns: list[str]) -> str:
    for pattern in patterns:
        match = re.search(pattern, user_agent, re.IGNORECASE)
        if match:
            return (match.group(1) or "").strip()
    return ""


def _guess_os_name(user_agent: str) -> str:
    ua = user_agent.lower()
    if "windows nt" in ua or "windows" in ua:
        return "Windows"
    if "iphone" in ua or "ipad" in ua or "ipod" in ua or "ios" in ua:
        return "iOS"
    if "android" in ua:
        return "Android"
    if "macintosh" in ua or "mac os x" in ua:
        return "macOS"
    if "cros" in ua:
        return "ChromeOS"
    if "linux" in ua:
        return "Linux"
    return "Unknown OS"


def _guess_browser_name_and_version(user_agent: str) -> tuple[str, str]:
    ua = user_agent.lower()

    if "edg/" in ua or "edge/" in ua or "edga/" in ua or "edgios/" in ua:
        return "Edge", _extract_version(user_agent, [r"(?:Edg|Edge|EdgA|EdgiOS)/([\d\.]+)"])
    if "opr/" in ua or "opera" in ua:
        return "Opera", _extract_version(user_agent, [r"(?:OPR|Opera)/([\d\.]+)"])
    if "samsungbrowser/" in ua:
        return "Samsung Internet", _extract_version(user_agent, [r"SamsungBrowser/([\d\.]+)"])
    if "firefox/" in ua or "fxios/" in ua:
        return "Firefox", _extract_version(user_agent, [r"(?:Firefox|FxiOS)/([\d\.]+)"])
    if "micromessenger/" in ua:
        return "WeChat", _extract_version(user_agent, [r"MicroMessenger/([\d\.]+)"])
    if "ucbrowser/" in ua:
        return "UC Browser", _extract_version(user_agent, [r"UCBrowser/([\d\.]+)"])
    if "qqbrowser/" in ua:
        return "QQ Browser", _extract_version(user_agent, [r"QQBrowser/([\d\.]+)"])
    if "msie " in ua or "trident/" in ua:
        return "Internet Explorer", _extract_version(user_agent, [r"MSIE\s([\d\.]+)", r"rv:([\d\.]+)"])

    # Safari must be checked before Chrome fallback when Version/... exists.
    if "safari/" in ua and "chrome/" not in ua and "crios/" not in ua:
        return "Safari", _extract_version(user_agent, [r"Version/([\d\.]+)"])

    if "chrome/" in ua or "crios/" in ua:
        return "Chrome", _extract_version(user_agent, [r"(?:Chrome|CriOS)/([\d\.]+)"])

    return "Unknown Browser", ""


def _guess_platform(os_name: str) -> str:
    if os_name == "iOS":
        return "ios"
    if os_name == "Android":
        return "android"
    if os_name in {"Windows", "macOS", "Linux", "ChromeOS"}:
        return "desktop"
    return "web"


def parse_user_agent(user_agent: str) -> Dict[str, str]:
    """Parse user agent into normalized os/browser/platform fields."""
    ua = str(user_agent or "").strip()
    os_name = _guess_os_name(ua)
    browser_name, browser_version = _guess_browser_name_and_version(ua)
    return {
        "os_name": os_name,
        "browser_name": browser_name,
        "browser_version": browser_version,
        "platform": _guess_platform(os_name),
    }


def build_device_label(os_name: str, browser_name: str) -> str:
    safe_os = str(os_name or "").strip() or "Unknown OS"
    safe_browser = str(browser_name or "").strip() or "Unknown Browser"
    return f"{safe_os} · {safe_browser}"


def is_generic_device_name(device_name: str) -> bool:
    value = str(device_name or "").strip()
    lowered = value.lower()
    if lowered in _GENERIC_DEVICE_NAMES:
        return True
    if re.match(r"^(desktop|web|android|ios|unknown)\s*[·\-]\s*[a-z]{2}(?:-[a-z]{2})?$", lowered):
        return True
    return lowered.endswith(" browser")


def normalize_device_name(device_name: str, parsed_ua: Dict[str, str]) -> str:
    """Return preferred device name; fallback to parsed OS+browser label."""
    raw = str(device_name or "").strip()
    if raw and not is_generic_device_name(raw):
        return raw[:120]
    return build_device_label(parsed_ua.get("os_name"), parsed_ua.get("browser_name"))[:120]
